File: 13. Find min max in BST.py
"""
Task: Find min and max in BST
Time Complexity: O(h)
Space Complexity: O(1)
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def find_min(root):
    if not root: raise ValueError('Empty')
    while root.left: root=root.left
    return root.data

def find_max(root):
    if not root: raise ValueError('Empty')
    while root.right: root=root.right
    return root.data

if __name__=='__main__':
    root=Node(5); root.left=Node(3); root.right=Node(7)
    print('Min:', find_min(root), 'Max:', find_max(root))

File: 13. Find min max in BST.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
int findMin(Node* root)
{
    if (!root)
        throw runtime_error("Empty");
    while (root->left) root = root->left;
    return root->data;
}
int findMax(Node* root)
{
    if (!root)
        throw runtime_error("Empty");
    while (root->right) root = root->right;
    return root->data;
}
int main()
{
    Node* root = new Node(5);
    root->left = new Node(3);
    root->right = new Node(7);
    cout << "Min: " << findMin(root) << " Max: " << findMax(root) << endl;
    return 0;
}
